woocommerce/google-listings-and-ads
# Pre-fill shipping rates in Setup MC

### Changes proposed in this Pull Request:

Based on #1160, depends on shipping rates suggestions developed by @nima-karimi in PR #1144. 

Prior to this PR, when users go through the Setup MC flow and arrive at Step 3 page, they would have to manually key in shipping rates settings, even though they already have something similar setup in WooCommerce settings.

In this PR, we use the shipping rates suggestions API and pre-fill it into the Setup MC Step 3 form. The logic goes like this: 

1. We call shipping rates API to check if there is existing shipping rates already setup, and also shipping rates suggestions API to get the suggestions.
2. Check shipping rates: 
    - If there are existing shipping rates, we will use the existing shipping rates, and ignore the shipping rates suggestions. It would behave like existing production behavior.
    - If there are no existing shipping rates, we will save the suggestions as shipping rates. When the saving is done, we display the UI with the values filled in.

**Technical notes:**

- I added a new hook called `useCallbackOnceEffect`. It allows us to call a callback function only once, when a condition is set to `true`.

### Screenshots:

📹 Demo video with voice:

https://user-images.githubusercontent.com/417342/146678326-4431bf2a-6dc8-4c8d-a880-c6005ff8d4e5.mov

### Detailed test instructions:

1. Make sure you have some shipping zones and shipping methods already setup in WooCommerce settings: `/wp-admin/admin.php?page=wc-settings&tab=shipping`. Make sure you have different rates, so that you will see multiple shipping rates rows in the following steps.
2. Go through Setup MC flow until you arrive at Step 3. Open browser dev tools network panel to inspect the API calls.
3. Click on the "I have a fairly simple shipping setup and I can estimate flat shipping rates and times." radio button. The estimated shipping rates section should show up with a loading indicator. 
4. While the loading indicator is displayed, there should be a number of API calls in the network panel. 
    1. `GET /mc/shipping/rates` and `GET /mc/shipping/rates/suggestions/batch` with country codes in the query string. 
    2. Because you do not have any existing shipping rates, the suggestions will be used and saved as your shipping rates. You should see multiple calls of `POST /mc/shipping/rates/batch` to save the suggestions as shipping rates.
5. The shipping rates countries and values should show up.
6. Delete or modify one of the shipping rates, and then reload the page. The shipping rates setup should be the same as prior to the reload. (Note: if you modify the shipping rates and then reload the page, you might encounter issue https://github.com/woocommerce/google-listings-and-ads/issues/1158.)
7. Remove all the shipping rates, and then reload the page. You should see the shipping rates being filled with the shipping rates suggestions again.


### Changelog entry

> Add - Pre-fill shipping rates in Setup Merchant Center flow based on store's shipping settings.
